Allow clamping of digit size options (#1070)
unicode-org/message-format-wg: spec/functions/number.md

If the value of a _digit size option_ does not evaluate as a non-negative integer,
or if the value exceeds any implementation-defined and option-specific upper or lower limit,

> the implementation will emit a _Bad Option Error_ and

Reviewer: Should the "will" here be a "MUST"?

If the value of a _digit size option_ does not evaluate as a non-negative integer,
or if the value exceeds any implementation-defined and option-specific upper or lower limit,
a _Bad Option_ error is emitted.
If the value exceeds an implementation-defined upper or lower limit, 

> the implementation MAY replace it with an implementation-defined value.

Reviewer 2: Does this happen _instead of_ the Bad Option error?

Is there somewhere else in the spec that says what happens if the provided option value is not in-range, whether it is dropped, etc? (Seems like there would be)

Author: > Does this happen instead of the Bad Option error?

No. The spec (on the previous line!) says a Bad Option is emitted, so one is emitted. This proposed addition allows the implementation to do something other than dropping the option on the floor (which would presumably result in the default value for that implementation being used). It also would have an effect on the _resolved value_ of the option.

The spec in general does not define behavior like this because of the layering in the design. What happens in a function is the function's business. Other functions might have similar or different behavior, depending on their needs. (Some options, when in error, leave the function unable to proceed.)

There are two places in the spec where something else could happen (on another layer).

One is during [option resolution](https://www.unicode.org/reports/tr35/dev/tr35-messageFormat.html#option-resolution), but that only results in a map of options to resolved values. That is where the text says that option resolution cannot fail (produce a fatal error). @macchiati and others have proposed allowing function descriptors and to be used to validate values _before_ function resolution. Such evaluation might result in the Bad Option occurring sooner.

The other is during [function resolution](https://www.unicode.org/reports/tr35/dev/tr35-messageFormat.html#function-resolution) during formatting. Effectively, "function resolution" is when the function is actually called. The function itself decides whether a Bad Option is a _failure_ or not. If it is a failure, the fallback representation is the result of formatting. Presumably if one clamped the value of a Digit Size Option, one would not emit a fallback representation (there would be no point in clamping the value, since the result is something like `{$foo}` not a formatted value such as `0.123456789012345678901`) making the Bad Option just a non-fatal error.

Reviewer 2: ok, so maybe a better resolution here is to just say somewhere that when a Bad Option error occurs, implementations may _either_ drop the option value _or_ substitute an implementation-defined replacement?

> a _Bad Option_ error is emitted.
> If the value exceeds an implementation-defined upper or lower limit, 
> the implementation MAY replace it with an implementation-defined value.

Reviewer 3: To resolve the ambiguity noted by @sffc in https://github.com/unicode-org/message-format-wg/pull/1070#discussion_r2043120996:
```suggestion
a _Bad Option_ error is emitted and
the implementation MAY ignore the option value or
replace it with an implementation-defined value.
```

Reviewer 3: One part of the change that I'm proposing here is allowing an implementation to behave the same way for a digit size option ending up with either `-1` or `0` as values for an option that has an implementation-defined minimum of `1`. Is there a reason why this should not be allowed?

Author: There is no reason that should not be allowed. I think my proposal allows that (the implementation can ignore the option **_or_** replace it)?

> If the value exceeds an implementation-defined upper or lower limit, 
> the implementation MAY replace the offending value with an implementation-defined value.

Reviewer 3: Nitpicky language suggestion, as "offending" seems a little surprising in this context.

```suggestion
If the value exceeds an implementation-defined upper or lower limit, 
the implementation MAY replace it with an implementation-defined value.
```
